File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/team/lobby/respawn/ExampleRespawnLobby.java
package com.gmail.borlandlp.minigamesdtools.arena.team.lobby.respawn;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaEventListener;
import com.gmail.borlandlp.minigamesdtools.arena.team.lobby.ArenaLobby;
import com.gmail.borlandlp.minigamesdtools.arena.team.lobby.PlayerLocker;
import org.bukkit.Bukkit;
import org.bukkit.boss.BarColor;
import org.bukkit.boss.BarStyle;
import org.bukkit.boss.BossBar;
import org.bukkit.entity.Player;

import java.time.Instant;
import java.util.*;

public class ExampleRespawnLobby extends ArenaLobby implements RespawnLobby, PlayerLocker {
    private int secondsWaiting;
    private Map<Player, Long> players = new Hashtable<>();// player -> unixtime
    private ArenaEventListener listener;
    private Map<Player, BossBar> bossBarMap = new Hashtable<>();

    @Override
    public void forceReleasePlayer(Player p) {
        this.removePlayer(p);
    }

    @Override
    public void addPlayer(Player player) {
        Debug.print(Debug.LEVEL.NOTICE, "Add player[name:" + player.getName() + "] to Respawn lobby");
        player.teleport(this.getSpawnPoint());
        this.players.put(player, Instant.now().getEpochSecond());

        BossBar bossBar = Bukkit.createBossBar("Возрождение", BarColor.RED, BarStyle.SOLID);
        bossBar.setProgress(1D);
        bossBar.setVisible(true);
        bossBar.addPlayer(player);
        this.bossBarMap.put(player, bossBar);

        if(MinigamesDTools.getInstance().getHotbarAPI().isBindedPlayer(player)) {
            MinigamesDTools.getInstance().getHotbarAPI().unbindHotbar(player);
        }
        if(this.isHotbarEnabled()) {
            MinigamesDTools.getInstance().getHotbarAPI().bindHotbar(this.getHotbarFor(player), player);
        }
    }

    @Override
    public void removePlayer(Player player) {
        this.bossBarMap.get(player).removePlayer(player);
        this.bossBarMap.remove(player);
        this.players.remove(player);

        if(this.isHotbarEnabled()) {
            MinigamesDTools.getInstance().getHotbarAPI().unbindHotbar(player);
        }
        if(this.getTeamProvider().getArena().getHotbarController().isEnabled()) {
            try {
                MinigamesDTools.getInstance().getHotbarAPI().bindHotbar(this.getTeamProvider().getArena().getHotbarController().buildDefaultHotbarFor(player), player);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onInit() {
        this.listener = new ExampleRespawnLobbyListener(this);
        try {
            this.getTeamProvider().getArena().getEventAnnouncer().register(this.listener);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void beforeGameStarting() {

    }

    @Override
    public void gameEnded() {
        for (Player player : new ArrayList<>(this.players.keySet())) {
            this.removePlayer(player);
        }
    }

    // draw inventorygui and other func
    @Override
    public void update() {
        for (Player player : this.bossBarMap.keySet()) {
            long remainS = (this.players.get(player) + this.secondsWaiting) - Instant.now().getEpochSecond();
            this.bossBarMap.get(player).setTitle("До возрождения: " + remainS + " с.");
            this.bossBarMap.get(player).setProgress((double)remainS / (double)this.secondsWaiting);
        }
    }

    @Override
    public void beforeRoundStarting() {

    }

    @Override
    public void onRoundEnd() {

    }

    @Override
    public Set<Player> getReadyPlayersToRespawn() {
        Set<Player> rPlayers = new HashSet<>();
        for (Player player : this.players.keySet()) {
            if((this.players.get(player) + this.secondsWaiting) <= Instant.now().getEpochSecond()) {
                rPlayers.add(player);
            }
        }

        return rPlayers;
    }

    @Override
    public Map<Player, Long> getWaitingPlayers() {
        return new Hashtable<>(this.players);
    }

    public void setSecondsWaiting(int secondsWaiting) {
        this.secondsWaiting = secondsWaiting;
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/team/lobby/respawn/ExampleRespawnLobby.kt
package com.gmail.borlandlp.minigamesdtools.arena.team.lobby.respawn

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools.Companion.instance
import com.gmail.borlandlp.minigamesdtools.arena.ArenaEventListener
import com.gmail.borlandlp.minigamesdtools.arena.team.lobby.ArenaLobby
import com.gmail.borlandlp.minigamesdtools.arena.team.lobby.PlayerLocker
import org.bukkit.Bukkit
import org.bukkit.boss.BarColor
import org.bukkit.boss.BarStyle
import org.bukkit.boss.BossBar
import org.bukkit.entity.Player
import java.time.Instant
import java.util.*

class ExampleRespawnLobby : ArenaLobby(), RespawnLobby, PlayerLocker {
    private var secondsWaiting = 0
    private val players: MutableMap<Player, Long> = Hashtable() // player -> unixtime
    private var listener: ArenaEventListener? = null
    private val bossBarMap: MutableMap<Player, BossBar> = Hashtable()

    override fun forceReleasePlayer(p: Player) {
        removePlayer(p)
    }

    override fun addPlayer(player: Player) {
        print(
            Debug.LEVEL.NOTICE,
            "Add player[name:" + player.name + "] to Respawn lobby"
        )
        player.teleport(spawnPoint)
        players[player] = Instant.now().epochSecond
        val bossBar =
            Bukkit.createBossBar("Возрождение", BarColor.RED, BarStyle.SOLID)
        bossBar.progress = 1.0
        bossBar.isVisible = true
        bossBar.addPlayer(player)
        bossBarMap[player] = bossBar
        if (instance!!.hotbarAPI!!.isBindedPlayer(player)) {
            instance!!.hotbarAPI!!.unbindHotbar(player)
        }
        if (this.isHotbarEnabled) {
            instance!!.hotbarAPI!!.bindHotbar(getHotbarFor(player)!!, player)
        }
    }

    override fun removePlayer(player: Player) {
        bossBarMap[player]!!.removePlayer(player)
        bossBarMap.remove(player)
        players.remove(player)
        if (this.isHotbarEnabled) {
            instance!!.hotbarAPI!!.unbindHotbar(player)
        }
        if (teamProvider!!.arena.hotbarController!!.isEnabled) {
            try {
                instance!!.hotbarAPI!!.bindHotbar(
                    teamProvider!!.arena.hotbarController!!.buildDefaultHotbarFor(
                        player
                    ), player
                )
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun onInit() {
        listener = ExampleRespawnLobbyListener(this)
        try {
            teamProvider!!.arena.eventAnnouncer.register(listener as ExampleRespawnLobbyListener)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun beforeGameStarting() {}
    override fun gameEnded() {
        for (player in ArrayList(players.keys)) {
            removePlayer(player)
        }
    }

    // draw inventorygui and other func
    override fun update() {
        for (player in bossBarMap.keys) {
            val remainS = players[player]!! + secondsWaiting - Instant.now().epochSecond
            bossBarMap[player]!!.title = "До возрождения: $remainS с."
            bossBarMap[player]!!.progress = remainS.toDouble() / secondsWaiting.toDouble()
        }
    }

    override fun beforeRoundStarting() {}
    override fun onRoundEnd() {}
    override val readyPlayersToRespawn: Set<Player>
        get() {
            val rPlayers: MutableSet<Player> = HashSet()
            for (player in players.keys) {
                if (players[player]!! + secondsWaiting <= Instant.now().epochSecond) {
                    rPlayers.add(player)
                }
            }
            return rPlayers
        }

    override val waitingPlayers: Map<Player, Long>
        get() = Hashtable(players)

    fun setSecondsWaiting(secondsWaiting: Int) {
        this.secondsWaiting = secondsWaiting
    }
}